Sources/SmithyCBOR/Writer/Writer.swift

```diff
@@ -0,0 +1,224 @@
+        guard let value else { return }
+        self.cborValue = .text(value)
+    }
+
+    public func write(_ value: Float?) throws {
+        guard let value else { return }
+        self.cborValue = .double(Double(value))
+    }
+
+    public func write(_ value: Double?) throws {
+        guard let value else { return }
+        self.cborValue = .double(value)
+    }
+
+    public func write(_ value: Int?) throws {
+        guard let value else { return }
+        self.cborValue = .int(Int64(value))
+    }
+
+    public func write(_ value: Int8?) throws {
+        guard let value else { return }
+        self.cborValue = .int(Int64(value))
+    }
+
+    public func write(_ value: Int16?) throws {
+        guard let value else { return }
+        self.cborValue = .int(Int64(value))
+    }
+
+    public func write(_ value: UInt8?) throws {
+        guard let value else { return }
+        self.cborValue = .uint(UInt64(value))
+    }
+
+    public func write(_ value: Data?) throws {
+        guard let value else { return }
+        self.cborValue = .bytes(value)
+    }
+
+    public func write(_ value: SmithyDocument?) throws {
+        // No operation.  Smithy document not supported in CBOR
+    }
+
+    public func writeTimestamp(_ value: Date?, format: TimestampFormat) throws {
+        guard let value else { return }
+        switch format {
+        case .epochSeconds:
+            self.cborValue = .date(value)
+        case .dateTime, .httpDate:
+            let string = TimestampFormatter(format: format).string(from: value)
+            self.cborValue = .text(string)
+        }
+    }
+
+    public func write<T>(_ value: T?) throws where T: RawRepresentable, T.RawValue == Int {
+        if let rawValue = value?.rawValue {
+            try write(rawValue)
+        }
+    }
+
+    public func write<T>(_ value: T?) throws where T: RawRepresentable, T.RawValue == String {
+        if let rawValue = value?.rawValue {
+            try write(rawValue)
+        }
+    }
+
+    public func writeMap<T>(
+        _ value: [String: T]?,
+        valueWritingClosure: (T, Writer) throws -> Void,
+        keyNodeInfo: NodeInfo,
+        valueNodeInfo: NodeInfo,
+        isFlattened: Bool
+    ) throws {
+        guard let value else { return }
+        var map: [String: CBORType] = [:]
+
+        for (key, val) in value {
+            let writer = self[key]
+            try valueWritingClosure(val, writer)
+
+            // If the writer itself doesn't have a cborValue, build it from its children
+            if writer.cborValue == nil, !writer.children.isEmpty {
+                var childMap: [String: CBORType] = [:]
+                for child in writer.children {
+                    if let childCborValue = child.cborValue {
+                        childMap[child.nodeInfo] = childCborValue
+                    }
+                }
+                writer.cborValue = .map(childMap) // Construct the map for the writer
```

Comment:
Q1: What happens if any of the children's cborValue wasn't yet assigned either? Or is that not possible?
Q2: What if a child map contains another nested child map? Doesn't this only handle one level of nesting?

Reply:
**Q1:** It is possible for a child’s cborValue to remain unassigned if you never write a value to it. However, that’s okay: if a child has children of its own, writeMap may build a .map(...) from those children, or the final encode(encoder:) pass will encode it as an indefinite-length map (see code block starting line 48). If the child has no children and no cborValue, it just gets skipped or encoded as an empty map.

**Q2:** Multiple nesting levels are supported. Even though writeMap appears to build just a one-level map, it considers each child to be a writer itself which allows the real nesting to occur during encode(encoder:), which recurses down the entire tree of child writers. This means a child map can contain another map, which can contain another map, and so on—no matter how deeply nested.